### PR TITLE
[Prototype] Add DependencyContext fallback to Razor compilation

### PR DESCRIPTION
**Pull request by pakrym**
Have a general look please:

@davidfowl @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc/Microsoft.AspNet.Mvc.xproj [hunk 1]
```diff
@@ -10,6 +10,9 @@
     <BaseIntermediateOutputPath Condition="'$(BaseIntermediateOutputPath)'=='' ">..\..\artifacts\obj\$(MSBuildProjectName)</BaseIntermediateOutputPath>
     <OutputPath Condition="'$(OutputPath)'=='' ">..\..\artifacts\bin\$(MSBuildProjectName)\</OutputPath>
   </PropertyGroup>
+  <PropertyGroup Label="Configuration">
+    <RootNamespace>Microsoft.AspNet.Mvc</RootNamespace>
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc/Microsoft.AspNet.Mvc.xproj]: ?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/DependencyContextCompilationService.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            IRazorViewEngineFileProviderAccessor fileProviderAccessor)
+            : base(environment, host, optionsAccessor, fileProviderAccessor)
+        {
+        }
+
+        protected override List<MetadataReference> GetApplicationReferences()
+        {
+            return DependencyContext.Default.CompileLibraries
+                .SelectMany(LibraryResolver.ResolvePaths)
+                .Select(this.CreateMetadataFileReference)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/DependencyContextCompilationService.cs]: nit: `this.`?

File: src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs [hunk 3]
```diff
@@ -119,7 +120,9 @@ public static IMvcCoreBuilder AddRazorViewEngine(this IMvcCoreBuilder builder)
         // Internal for testing.
         internal static void AddRazorViewEngineServices(IServiceCollection services)
         {
-            if (CompilationServices.Default != null)
+            var isDotnet = CompilationServices.Default == null;
+
+            if (!isDotnet)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs]: Commonize it under one if-else.

File: src/Microsoft.AspNet.Mvc.Razor/DependencyContextRazorViewEngineOptionsSetup.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNet.Mvc.Razor;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.DependencyModel;
+
+namespace HelloMvc
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/DependencyContextRazorViewEngineOptionsSetup.cs]: name

**Comment by pranavkm** [comment]
Prototype looks pretty good. Does it run with Dorkbuild?

File: src/Microsoft.AspNet.Mvc.Razor/DependencyContextRazorViewEngineOptionsSetup.cs (hunk 4 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/DependencyContextRazorViewEngineOptionsSetup.cs]: What :smile:

File: src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs [hunk 5]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Linq;
 using System.Reflection;
+using HelloMvc;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs]: :smile:

File: src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs [hunk 6]
```diff
@@ -119,7 +120,9 @@ public static IMvcCoreBuilder AddRazorViewEngine(this IMvcCoreBuilder builder)
         // Internal for testing.
         internal static void AddRazorViewEngineServices(IServiceCollection services)
         {
-            if (CompilationServices.Default != null)
+            var isDotnet = CompilationServices.Default == null;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs]: change the variable name to be less about dotnet and more about compilation services missing.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs [hunk 7]
```diff
@@ -21,20 +21,18 @@
...
 using Microsoft.Extensions.Options;
-using Microsoft.AspNet.Diagnostics;
 
 namespace Microsoft.AspNet.Mvc.Razor.Compilation
 {
     /// <summary>
     /// A type that uses Roslyn to compile C# content.
     /// </summary>
-    public class RoslynCompilationService : ICompilationService
+    public abstract class RoslynCompilationService : ICompilationService
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: If one of the derived types is meant to be throw-away code, I'd rather we duplicate this so we don't need to deal with this when we eventually remove the DNX support.

Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: Meh, it's a change anyways and @pakrym loves inheritance :smile:

**Comment by davidfowl** [comment]
looks good.

**Comment by pakrym** [comment]
Yes, with latest dotnet cli fixes and this pr it works

File: NuGet.config [hunk 8]
```diff
@@ -1,6 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <packageSources>
+    <add key="dotnet-cli" value="https://www.myget.org/F/dotnet-cli/api/v3/index.json" />
```
Comment (pranavkm) [inline NuGet.config]: Do we need this for the `DependencyContext` package? We should mirror these as part of the CoreCLR job so we can gate any breaking behavior.

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/DependencyContextAssemblyProvider.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNet.Mvc.Infrastructure
+{
+    public class DependencyContextAssemblyProvider: IAssemblyProvider
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/DependencyContextAssemblyProvider.cs]: Formatting (spaces)

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/DependencyContextAssemblyProvider.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            "Microsoft.AspNet.Mvc.Cors",
+            "Microsoft.AspNet.Mvc.DataAnnotations",
+            "Microsoft.AspNet.Mvc.Formatters.Json",
+            "Microsoft.AspNet.Mvc.Formatters.Xml",
+            "Microsoft.AspNet.Mvc.Localization",
+            "Microsoft.AspNet.Mvc.Razor",
+            "Microsoft.AspNet.Mvc.Razor.Host",
+            "Microsoft.AspNet.Mvc.TagHelpers",
+            "Microsoft.AspNet.Mvc.ViewFeatures",
+            "Microsoft.AspNet.PageExecutionInstrumentation.Interfaces",
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/DependencyContextAssemblyProvider.cs]: Could you remove this? The package got removed but I missed removing it in the `DefaultAssemblyProvider`

**Comment by pranavkm** [comment]
I can mirror the packages from `dotnet-cli` feed as a one time job and figure out how to pull their packages in on a continuous basis. 

:shipit: once changes are addressed.

File: NuGet.config (hunk 8 above)
Comment (pakrym) [inline NuGet.config]: Yes, DependencyModel package is there

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/DependencyContextAssemblyProvider.cs (hunk 9 above)
Comment (pakrym) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/DependencyContextAssemblyProvider.cs]: ?

**Pull request merged**